fdbclient/MultiVersionTransaction.actor.cpp

```diff
@@ -588,7 +588,20 @@ ThreadFuture<Void> MultiVersionTransaction::onError(Error const& e) {
 	else {
 		auto tr = getTransaction();
 		auto f = tr.transaction ? tr.transaction->onError(e) : ThreadFuture<Void>(Never());
-		return abortableFuture(f, tr.onChange);
+		f = abortableFuture(f, tr.onChange);
+
+		return flatMapThreadFuture<Void, Void>(f, [this, e](ErrorOr<Void> ready) {
+			if(!ready.isError() || ready.getError().code() != error_code_cluster_version_changed) {
```

Comment:
Okay, so the exact problem here is:

1. The user sets up two client libraries and then uses one of them to connect to the cluster.
1. They do a thing that results in an error (like, say, get a transaction conflict).
1. While waiting on the error, the cluster is upgraded.
1. The `onError` call used to propagate that error up, but now it calls `onError` in the new thing and updates the transaction to use the new client library.

I think that means that for existing users of the retry loops, they would have just seen a retriable error propagate up rather than, say, seeing the transaction retry loop try again with an incompatible protocol version (right?). I ask because I think if it were the latter, then there would be a somewhat compelling case that this should go onto 6.1 as a patch, but I think it's fine on master if only the former.

Reply:
Yeah, that's a good summary of the problem.

The old behavior should have resulted in it bubbling up a retryable error, but since it came from `onError` it's likely they wouldn't have actually attempted to retry it.